backend/src/lib.rs

```diff
@@ -184,91 +72,51 @@ pub type Proof = Vec<u8>;
+        pil: &'a Analyzed<F>,
+        fixed: &'a [(String, Vec<F>)],
+        output_dir: Option<&'a Path>,
+        setup: Option<&mut dyn io::Read>,
+        verification_key: Option<&mut dyn io::Read>,
+    ) -> Result<Box<dyn Backend<'a, F> + 'a>, Error>;
+
+    /// Generate a new setup.
+    fn generate_setup(&self, _size: DegreeType, _output: &mut dyn io::Write) -> Result<(), Error> {
+        Err(Error::NoSetupAvailable)
+    }
+}
 
+/// Dynamic interface for a backend.
+pub trait Backend<'a, F: FieldElement> {
     /// Perform the proving.
     ///
     /// If prev_proof is provided, proof aggregation is performed.
     ///
-    /// Returns the generated proof, and the string serialization of the
-    /// constraints.
+    /// Returns the generated proof.
     fn prove(
         &self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
         witness: &[(String, Vec<F>)],
         prev_proof: Option<Proof>,
-    ) -> Result<(Proof, Option<String>), Error>;
+    ) -> Result<Proof, Error>;
 
-    /// Write the prover setup to a file, so that it can be loaded later.
-    fn write_setup(&self, output: &mut dyn io::Write) -> Result<(), Error>;
-}
-
-/// Dynamic interface for a backend factory.
-pub trait BackendFactory<F: FieldElement> {
-    /// Maybe perform the setup, and create a new backend object.
-    fn create(&self, degree: DegreeType) -> Box<dyn Backend<F>>;
-
-    /// Create a backend object from a prover setup loaded from a file.
-    fn create_from_setup(&self, input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error>;
-}
-
-/*
-    Below are the traits implemented by the backends.
-*/
-
-/// Trait implemented by all backends.
-trait BackendImpl<F: FieldElement> {
-    fn new(degree: DegreeType) -> Self;
-
-    fn add_verification_key(
-        &mut self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-        vkey: Vec<u8>,
-    );
-
-    fn verification_key(
-        &self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-    ) -> Result<Vec<u8>, Error>;
-
-    fn verify(&self, proof: &Proof, instances: &[Vec<F>]) -> Result<(), Error>;
-
-    fn prove(
-        &self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-        witness: &[(String, Vec<F>)],
-        prev_proof: Option<Proof>,
-    ) -> Result<(Proof, Option<String>), Error>;
-}
+    /// Verifies a proof.
+    fn verify(&self, _proof: &Proof, _instances: &[Vec<F>]) -> Result<(), Error> {
+        Err(Error::NoVerificationAvailable)
+    }
 
-/// Trait implemented by backends that have a setup phase that must be saved to
-/// a file.
-trait BackendImplWithSetup<F: FieldElement>
-where
-    Self: Sized + BackendImpl<F>,
-{
-    /// Create a backend object from a setup loaded from a file.
-    fn new_from_setup(input: &mut dyn io::Read) -> Result<Self, io::Error>;
+    /// Exports the setup in a backend specific format. Can be used to create
```

Comment:
wdym "can be used to create a new instance of the same backend"?

Reply:
"can be used to create a new instance of the same backend type" makes it clearer? Means that I can pass the exported setup to `create()` function.

Reply:
reworded